### PR TITLE
address potential strict aliasing violation

### PR DESCRIPTION
**Pull request by veorq**

### CODE REVIEW

File: halfsiphash.c [hunk 1]
```diff
@@ -69,19 +69,22 @@
 #define TRACE
 #endif
 
-int halfsiphash(const uint8_t *in, const size_t inlen, const uint8_t *k,
-                uint8_t *out, const size_t outlen) {
+int halfsiphash(const void *in, const size_t inlen, const void *k, uint8_t *out,
+                const size_t outlen) {
+
+    unsigned char *ni = (unsigned char *)in;
+    unsigned char *kk = (unsigned char *)k;
```
Comment (gsauthof) [inline halfsiphash.c:76]: You cast away the `const` qualifier which conflicts with the role of that pointer.

So what about:

```
    const unsigned char *ni = (const unsigned char *)in;
￼    const unsigned char *kk = (const unsigned char *)k;
```

For C compiler only compatibility, you could even leave out the cast on the right hand side.

File: halfsiphash.c [hunk 2]
```diff
@@ -69,19 +69,22 @@
...
     uint32_t v2 = UINT32_C(0x6c796765);
     uint32_t v3 = UINT32_C(0x74656462);
-    uint32_t k0 = U8TO32_LE(k);
-    uint32_t k1 = U8TO32_LE(k + 4);
+    uint32_t k0 = U8TO32_LE(kk);
+    uint32_t k1 = U8TO32_LE(kk + 4);
     uint32_t m;
     int i;
-    const uint8_t *end = in + inlen - (inlen % sizeof(uint32_t));
+    const uint8_t *end = ni + inlen - (inlen % sizeof(uint32_t));
```
Comment (gsauthof) [inline halfsiphash.c:87]: This should also be:

```
const unsigned char *end = ni + inlen - (inlen % sizeof(uint32_t));
```

File: siphash.c [hunk 3]
```diff
@@ -76,19 +76,22 @@
 #define TRACE
 #endif
 
-int siphash(const uint8_t *in, const size_t inlen, const uint8_t *k,
-            uint8_t *out, const size_t outlen) {
+int siphash(const void *in, const size_t inlen, const void *k, uint8_t *out,
+            const size_t outlen) {
+
+    unsigned char *ni = (unsigned char *)in;
+    unsigned char *kk = (unsigned char *)k;
```
Comment (gsauthof) [inline siphash.c:83]: As above, `const unsigned char *ni ... const unsigned char *kk ...`, etc.

File: siphash.c [hunk 4]
```diff
@@ -76,19 +76,22 @@
...
     uint64_t v2 = UINT64_C(0x6c7967656e657261);
     uint64_t v3 = UINT64_C(0x7465646279746573);
-    uint64_t k0 = U8TO64_LE(k);
-    uint64_t k1 = U8TO64_LE(k + 8);
+    uint64_t k0 = U8TO64_LE(kk);
+    uint64_t k1 = U8TO64_LE(kk + 8);
     uint64_t m;
     int i;
-    const uint8_t *end = in + inlen - (inlen % sizeof(uint64_t));
+    const uint8_t *end = ni + inlen - (inlen % sizeof(uint64_t));
```
Comment (gsauthof) [inline siphash.c:94]: As above, `const unsigned char *end = ...`.

**Review by gsauthof — CHANGES_REQUESTED** [review]

**Comment by veorq** [comment]
Thanks @gsauthof for catching these, no fixed, we should be good to go!

**Pull request merged**